File: modules/zheng-background/ios/ZhengBackgroundModule.swift
import ExpoModulesCore
import UIKit

public class ZhengBackgroundModule: Module {
  private var backgroundTaskId: UIBackgroundTaskIdentifier = .invalid
  private var backgroundRefCount = 0

  public func definition() -> ModuleDefinition {
    Name("ZhengBackground")

    AsyncFunction("moveToBackground") {
      await MainActor.run {
        UIApplication.shared.perform(#selector(NSXPCConnection.suspend))
      }
    }

    AsyncFunction("beginBackgroundExecution") {
      await MainActor.run {
        self.incrementBackgroundTask()
      }
    }

    AsyncFunction("endBackgroundExecution") {
      await MainActor.run {
        self.decrementBackgroundTask()
      }
    }
  }

  private func incrementBackgroundTask() {
    backgroundRefCount += 1
    guard backgroundTaskId == .invalid else { return }
    backgroundTaskId = UIApplication.shared.beginBackgroundTask(withName: "ZhengAutoLedger") {
      self.backgroundRefCount = 0
      self.finishBackgroundTask()
    }
  }

  private func decrementBackgroundTask() {
    backgroundRefCount = max(0, backgroundRefCount - 1)
    if backgroundRefCount == 0 {
      finishBackgroundTask()
    }
  }

  private func finishBackgroundTask() {
    guard backgroundTaskId != .invalid else { return }
    UIApplication.shared.endBackgroundTask(backgroundTaskId)
    backgroundTaskId = .invalid
  }
}

File: modules/zheng-background/android/src/main/java/expo/modules/zhengbackground/ZhengBackgroundModule.kt
package expo.modules.zhengbackground

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ZhengBackgroundModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ZhengBackground")

    AsyncFunction("moveToBackground") {
      val activity = appContext.currentActivity
      activity?.moveTaskToBack(true)
    }

    AsyncFunction("beginBackgroundExecution") {
      val context = appContext.reactContext ?: return@AsyncFunction
      ZhengBackgroundForegroundService.begin(context)
    }

    AsyncFunction("endBackgroundExecution") {
      val context = appContext.reactContext ?: return@AsyncFunction
      ZhengBackgroundForegroundService.end(context)
    }
  }
}
